File: lab04.1/ciphers_python/playfer_cipher.py
CYRILLIC_ALPHABET = "АБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
LATIN_ALPHABET = "ABCDEFGHIKLMNOPQRSTUVWXYZ"


def _get_table_chars(*, keyword: str, alphabet: str) -> list[str]:
    table_chars = []
    for ch in (keyword + alphabet):
        if ch not in table_chars:
            table_chars.append(ch)
    return table_chars


def _make_table(*, keyword: str, alphabet: str, table_size: int | None = None) -> list[list[str]]:
    table_chars = _get_table_chars(keyword=keyword, alphabet=alphabet)
    if table_size is None:
        table_size = len(set(keyword))
    table = [[]]
    for ch in table_chars:
        if len(table[-1]) >= table_size:
            table.append([])
        table[-1].append(ch)
    return table


def _find_table_index(ch: str, table: list[list[str]]) -> int | tuple[int, int]:
    for i in range(len(table)):
        for j in range(len(table[i])):
            if table[i][j] == ch:
                return (i, j)
    return -1


def _get_digrams(message: str, delimeter: str, alphabet: str) -> str:
    digram_letters = 0
    last_letter = ''
    digrams = ""
    for ch in message.upper():
        if ch not in alphabet.upper():
            continue
        if digram_letters >= 2:
            digrams += ' '
            digram_letters = 0
        if digram_letters == 1 and ch == last_letter:
            digrams += delimeter + ' '
            digram_letters = 0
        digrams += ch
        last_letter = ch
        digram_letters += 1
    if digram_letters == 1:
        digrams += delimeter
    return digrams


def _normalize_index(index: tuple[int, int], table: list[list[str]]) -> tuple[int, int]:
    new_index = [index[0], index[1]]
    if new_index[0] >= len(table):
        new_index[0] = 0
    elif new_index[0] < 0:
        new_index[0] = len(table) - 1
    if new_index[1] >= len(table[new_index[0]]):
        new_index[1] = 0
    elif new_index[1] < 0:
        new_index[1] = len(table[new_index[0]]) - 1
    return (new_index[0], new_index[1])


def _playfer_cipher(*, old_message: str, keyword: str, alphabet: str, help_char: str, step: int, table_size: int) -> str:
    table = _make_table(
        keyword=keyword.upper(),
        alphabet=alphabet.upper(),
        table_size=table_size,
    )
    step = int(step / abs(step))
    old_digrams = _get_digrams(old_message, help_char, alphabet)
    new_digrams = ""
    for old_digram in old_digrams.split(' '):
        old_ch_1_index = _find_table_index(old_digram[0], table)
        old_ch_2_index = _find_table_index(old_digram[1], table)
        if old_ch_1_index[0] == old_ch_2_index[0]:
            new_ch_1_index = (old_ch_1_index[0], old_ch_1_index[1] + step)
            new_ch_2_index = (old_ch_2_index[0], old_ch_2_index[1] + step)
        elif old_ch_1_index[1] == old_ch_2_index[1]:
            new_ch_1_index = (old_ch_1_index[0] + step, old_ch_1_index[1])
            new_ch_2_index = (old_ch_2_index[0] + step, old_ch_2_index[1])
        else:
            new_ch_1_index = (old_ch_1_index[0], old_ch_2_index[1])
            new_ch_2_index = (old_ch_2_index[0], old_ch_1_index[1])
        new_ch_1_index = _normalize_index(new_ch_1_index, table)
        new_ch_2_index = _normalize_index(new_ch_2_index, table)
        new_digram = table[new_ch_1_index[0]][new_ch_1_index[1]] + table[new_ch_2_index[0]][new_ch_2_index[1]]
        new_digrams += ' ' + new_digram
    return new_digrams.replace(' ', '')


def playfer_cipher_encryption(*, decrypted_message: str, keyword: str, alphabet: str, help_char: str, table_size: int) -> str:
    return _playfer_cipher(
        old_message=decrypted_message,
        keyword=keyword,
        alphabet=alphabet,
        help_char=help_char,
        step=1,
        table_size=table_size,
    )


def playfer_cipher_decryption(*, encrypted_message: str, keyword: str, alphabet: str, help_char: str, table_size: int) -> str:
    return _playfer_cipher(
        old_message=encrypted_message,
        keyword=keyword,
        alphabet=alphabet,
        help_char=help_char,
        step=-1,
        table_size=table_size,
    )


def playfer_cipher_cyrillic_encryption(decrypted_message: str, keyword: str) -> str:
    return playfer_cipher_encryption(
        decrypted_message=decrypted_message.replace('ё', 'е').replace('Ё', 'Е'),
        keyword=keyword,
        alphabet=CYRILLIC_ALPHABET,
        help_char='Я',
        table_size=4,
    )


def playfer_cipher_cyrillic_decryption(encrypted_message: str, keyword: str) -> str:
    return playfer_cipher_decryption(
        encrypted_message=encrypted_message,
        keyword=keyword,
        alphabet=CYRILLIC_ALPHABET,
        help_char='Я',
        table_size=4,
    )


def playfer_cipher_latin_encryption(decrypted_message: str, keyword: str) -> str:
    return playfer_cipher_encryption(
        decrypted_message=decrypted_message.replace('j', 'i').replace('J', 'I'),
        keyword=keyword,
        alphabet=LATIN_ALPHABET,
        help_char='X',
        table_size=5,
    )


def playfer_cipher_latin_decryption(encrypted_message: str, keyword: str) -> str:
    return playfer_cipher_decryption(
        encrypted_message=encrypted_message,
        keyword=keyword,
        alphabet=LATIN_ALPHABET,
        help_char='X',
        table_size=5,
    )

File: lab04.1/ciphers_csharp/PlayferCipher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ciphers_csharp {
    internal static class PlayferCipher {
        private const string CYRILLIC_ALPHABET = "АБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ";
        private const string LATIN_ALPHABET = "ABCDEFGHIKLMNOPQRSTUVWXYZ";

        private static List<char> _getTableChars(string keyword, string alphabet) {
            List<char> tableChars = new List<char>();
            foreach (char ch in (keyword + alphabet)) {
                if (!tableChars.Contains(ch)) {
                    tableChars.Add(ch);
                }
            }
            return tableChars;
        }

        private static List<List<char>> _makeTable(string keyword, string alphabet, int? tableSize = null) {
            List<char> tableChars = _getTableChars(keyword, alphabet);
            if (tableSize == null) {
                tableSize = keyword.Distinct().ToList().Count;
            }
            List<List<char>> table = new List<List<char>>() { new List<char>() };
            foreach (char ch in tableChars) {
                if (table.Last().Count >= tableSize) {
                    table.Add(new List<char>());
                }
                table.Last().Add(ch);
            }
            return table;
        }

        private static (int, int) _findTableIndex(char ch, List<List<char>> table) {
            for (int i = 0; i < table.Count; i++) {
                for (int j = 0; j < table[i].Count; j++) {
                    if (table[i][j] == ch) {
                        return (i, j);
                    }
                }
            }
            return (-1, -1);
        }

        private static string _getDigrams(string message, char delimeter, string alphabet) {
            int digramLetters = 0;
            char lastLetter = '\0';
            string digrams = "";
            foreach (char ch in message.ToUpper()) {
                if (!alphabet.ToUpper().Contains(ch)) {
                    continue;
                }
                if (digramLetters >= 2) {
                    digrams += ' ';
                    digramLetters = 0;
                }
                if (digramLetters == 1 && ch == lastLetter) {
                    digrams += delimeter + ' ';
                    digramLetters = 0;
                }
                digrams += ch;
                lastLetter = ch;
                digramLetters++;
            }
            if (digramLetters == 1) {
                digrams += delimeter;
            }
            return digrams;
        }

        private static (int, int) _normalizeIndex((int, int) index, List<List<char>> table) {
            (int, int) newIndex = (index.Item1, index.Item2);
            if (newIndex.Item1 >= table.Count) {
                newIndex.Item1 = 0;
            }
            else if (newIndex.Item1 < 0) {
                newIndex.Item1 = table.Count - 1;
            }
            if (newIndex.Item2 >= table[newIndex.Item1].Count) {
                newIndex.Item2 = 0;
            }
            else if (newIndex.Item2 < 0) {
                newIndex.Item2 = table[newIndex.Item1].Count - 1;
            }
            return newIndex;
        }

        private static string _playferCipher(string oldMessage, string keyword, string alphabet, char helpChar, int step, int tableSize) {
            List<List<char>> table = _makeTable(
                keyword.ToUpper(),
                alphabet.ToUpper(),
                tableSize
            );
            step = (int)(step / Math.Abs(step));
            string oldDigrams = _getDigrams(oldMessage, helpChar, alphabet);
            string newDigrams = "";
            foreach (string oldDigram in oldDigrams.Split(' ')) {
                (int, int) oldCh1Index = _findTableIndex(oldDigram[0], table);
                (int, int) oldCh2Index = _findTableIndex(oldDigram[1], table);
                (int, int) newCh1Index;
                (int, int) newCh2Index;
                if (oldCh1Index.Item1 == oldCh2Index.Item1) {
                    newCh1Index = (oldCh1Index.Item1, oldCh1Index.Item2 + step);
                    newCh2Index = (oldCh2Index.Item1, oldCh2Index.Item2 + step);
                }
                else if (oldCh1Index.Item2 == oldCh2Index.Item2) {
                    newCh1Index = (oldCh1Index.Item1 + step, oldCh1Index.Item2);
                    newCh2Index = (oldCh2Index.Item1 + step, oldCh2Index.Item2);
                }
                else {
                    newCh1Index = (oldCh1Index.Item1, oldCh2Index.Item2);
                    newCh2Index = (oldCh2Index.Item1, oldCh1Index.Item2);
                }
                newCh1Index = _normalizeIndex(newCh1Index, table);
                newCh2Index = _normalizeIndex(newCh2Index, table);
                string newDigram = string.Concat(table[newCh1Index.Item1][newCh1Index.Item2], table[newCh2Index.Item1][newCh2Index.Item2]);
                newDigrams += ' ' + newDigram;
            }
            return newDigrams.Replace(" ", "");
        }

        public static string PlayferCipherEncryption(string decryptedMessage, string keyword, string alphabet, char helpChar, int tableSize) {
            return _playferCipher(
                decryptedMessage,
                keyword,
                alphabet,
                helpChar,
                1,
                tableSize
            );
        }

        public static string PlayferCipherDecryption(string encryptedMessage, string keyword, string alphabet, char helpChar, int tableSize) {
            return _playferCipher(
                encryptedMessage,
                keyword,
                alphabet,
                helpChar,
                -1,
                tableSize
            );
        }

        public static string PlayferCipherCyrillicEncryption(string decryptedMessage, string keyword) {
            return PlayferCipherEncryption(
                decryptedMessage.Replace('ё', 'е').Replace('Ё', 'Е'),
                keyword,
                CYRILLIC_ALPHABET,
                'Я',
                4
            );
        }

        public static string PlayferCipherCyrillicDecryption(string encryptedMessage, string keyword) {
            return PlayferCipherDecryption(
                encryptedMessage,
                keyword,
                CYRILLIC_ALPHABET,
                'Я',
                4
            );
        }

        public static string PlayferCipherLatinEncryption(string decryptedMessage, string keyword) {
            return PlayferCipherEncryption(
                decryptedMessage.Replace('j', 'i').Replace('J', 'I'),
                keyword,
                LATIN_ALPHABET,
                'X',
                5
            );
        }

        public static string PlayferCipherLatinDecryption(string encryptedMessage, string keyword) {
            return PlayferCipherDecryption(
                encryptedMessage,
                keyword,
                LATIN_ALPHABET,
                'X',
                5
            );
        }
    }
}
